### PR TITLE
Draft Collator Proposal.

### PR DESCRIPTION
**Pull request by cschuet**
[Rendered](https://github.com/cschuet/rfcs/blob/collator-interface/text/0000-collator-interface.md)

### CODE REVIEW

File: text/0000-collator-interface.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+For cloud-based mapping we have to deal with the situation where the robot to cloud uplink is unrealible.
+A robot might venture into an area where no Wi-Fi signal is available thereby severing the cloud uplink.
+The collator in the Cartographer cloud instance would currently stall as it is waiting for data from that robot before allowing any other trajectories to make progress.
+Ideally this behavior would be configurable through `MapBuilderOptions`.
+
+## Approach
+[approach]: #approach
+
+We propose to introduce a `sensor::CollatorInterface` which would be a pure virtualization of the current `Collator`.
+The current `Collator` would be renamed to `AcrossTrajectoryCollator` and be marked as an implementation of that interface.
```
Comment (cschuet) [inline text/0000-collator-interface.md]: If anyone has a better names for the implementations I'd be grateful.

File: text/0000-collator-interface.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+A new `Collator` would be introduce, `PerTrajectoryCollator`, which only collates sensor data belonging to the same trajectory.
```
Comment (pifon2a) [inline text/0000-collator-interface.md]: nit: "introduced"

File: text/0000-collator-interface.md (hunk 1 above)
Comment (pifon2a) [reply text/0000-collator-interface.md]: Will we need 'AcrossTrajectoryCollator' at all? Is it possible to just replace the current Collator with 'PerTrajectoryCollator'? Or implement the approach you suggested and then compare results and drop 'AcrossTrajectoryCollator' if we want it to be done safely.

**Review by pifon2a — APPROVED** [review]

**Comment by SirVer** [comment]
@cschuet The rendered link should contain the name of the branch, not a particular SHA so that it always is current to the latest in this PR. I fixed this for you here.

**Comment by SirVer** [comment]
lgtm in general. Adding @wohe, maybe he has feelings about this.

**Review by SirVer — APPROVED** [review]

**Comment by wohe** [comment]
I think the RFC should mention that per-trajectory collating means that there no longer is a deterministic result given the same input data. Or do I misunderstand something here?

**Comment by cschuet** [comment]
@wohe I think you are right about that.
@pifon2a I think this is also the answer to your question: we want to retain the old trajectory builder to keep the possibility of having a deterministic result.

**Comment by gaschler** [comment]
LGTM. I also agree with wohe that the non-determinism of PerTrajectoryCollator should be documented.

**Review by gaschler — APPROVED** [review]

**Pull request merged**